### ISSUE TITLE
Possible discrepancy in GRPO loss: Paper vs. implementation (log-prob vs. prob)

### ISSUE THREAD

**Issue by liranringel** [huggingface/trl#2752]
In the last page of the [DeepSeekMath paper](https://arxiv.org/pdf/2402.03300), the authors define the gradient of the GRPO objective as follows:

![Image](https://github.com/user-attachments/assets/d0c72a85-d436-41a5-9294-4276becaccc4)

This formulation is a weighted sum of the log probabilities.

However, in the GRPO trainer, the loss is computed using the probabilities themselves instead of log probabilities, as indicated by the use of `torch.exp`:

![Image](https://github.com/user-attachments/assets/01cc57b1-9224-4fd3-900c-a58311a1beef)

Could you clarify the reason for this difference?

To better understand this, I derived what seems to be the implementation’s version of the gradient (note that the probability of the expectation does not depend on $\theta$):

![Image](https://github.com/user-attachments/assets/cac00ea9-d66c-4ea9-bf33-f8bfb6c801b7)

Should the implementation be using log probabilities to align with the paper's derivation, or is there a mistake in the paper?

Thanks!

**Comment by xuefei1** [comment]
I had the same question. I also specifically wonder about this line, why it's `per_token_logps - per_token_logps.detach()` inside `torch.exp`, shouldn't it be `ref_per_token_logps`?

![Image](https://github.com/user-attachments/assets/1f0b9f35-cbcd-440e-b381-16ceb9f8d91d)

Original line: https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L521

**Comment by cfpark00** [comment]
@liranringel 

I'm unsure about the gradient coefficient form, but the implementation seems consistent (well, in terms of log(p) vs. p) with the following? (From deepseekmath paper)

<img width="856" alt="Image" src="https://github.com/user-attachments/assets/823f72fe-821b-4cd0-a686-c3773e87f835" />

----

Then of course there is no clipping here. And as @xuefei1  says,

it seems like the current implementation is equivalent to pushing everything into the innermost loop of:

<img width="848" alt="Image" src="https://github.com/user-attachments/assets/f6e4bfd0-082c-49a9-9c71-cfce5e265acd" />

which I guess will make gradients tiny?

**Comment by liranringel** [comment]
@cfpark00 But they explicitly state that the policy update maximizes the GRPO objective as per Equation 21, which is the weighted log probability version. This suggests that the theoretical formulation is indeed meant to use log probabilities, yet the implementation appears to use probabilities instead.

Also, I'm still confused about Equation 20—it's unclear how they derived it. A derivation similar to policy gradients doesn't seem to work here, since the probability inside the expectation does not depend on $\theta$.

**Comment by qgallouedec** [comment]
I think there's a misunderstanding. In the paper it is written 

> "Update the policy model no by maximizing the GRPO objective (Equation 21)".

Equation 21 gives the gradient of the objective, but what we're trying to optimize is the objective, not the gradient. This is why we use equation (3) in practice, and autograd takes care of the derivation and optimization step.

**Comment by liranringel** [comment]
Well, the point is that in RL, the loss is not always just the (minus) expression inside the expectation.
If the expectation is taken over a distribution that depends on the model parameters, we have to account for that—typically using the log trick (see [this explanation from ChatGPT](https://chatgpt.com/share/67a39058-a6f8-8013-8987-0f9af8ebef36)).

Thus, I suspect that the DeepSeek researchers actually treated the distribution inside the expectation as parameterized by $\pi_\theta$, which allowed them to apply the log trick. I suspect that this might be the reason the logarithm appears in equation 20, though I am not certain.

To clarify, I believe that the gradient computed using autograd in the current implementation differs from the gradient in Equation 20 of the paper.

**Comment by cfpark00** [comment]
I see.. I'm too much of a beginner in RL to provide a real opinion here. However, I can point you towards the corresponding lines in:

open-instruct:
https://github.com/allenai/open-instruct/blob/6b3964bcbc819299ca1269b3c306cc498355e488/open_instruct/grpo_vllm_thread_ray_gtrl.py#L1205

verl:
https://github.com/volcengine/verl/blob/6872dbefd5258643db247febf7ca68748d13e09c/verl/trainer/ppo/core_algos.py#L185

openrlhf:
https://github.com/OpenRLHF/OpenRLHF/blob/631b0bcb7c14ec2fd4117a43a661903ed60d26cc/openrlhf/models/loss.py#L72


It seems like they are all implementing the same loss as TRL?

@qgallouedec  would it be worth tagging the devs of each library for a discussion or not?

**Comment by liranringel** [comment]
@cfpark00 @qgallouedec I continued my derivation using the log trick and can confirm that your implementation matches Equation 20, so I am closing this issue.

Thanks for the help!

Here’s the derivation:

![Image](https://github.com/user-attachments/assets/a00c1e49-8bde-4315-b50e-ed74f9ce4b43)

**Comment by Paulmzr** [comment]
> [@cfpark00](https://github.com/cfpark00) [@qgallouedec](https://github.com/qgallouedec) I continued my derivation using the log trick and can confirm that your implementation matches Equation 20, so I am closing this issue.

> Thanks for the help!
> 
> Here’s the derivation:
> 
> ![Image](https://github.com/user-attachments/assets/a00c1e49-8bde-4315-b50e-ed74f9ce4b43)

The derivation is NOT right. When you try to estimate KL using q/p - log(q/p) - 1, you need to sample from p rather than q. So the sum operation in terms of KL should be on samples from pi_theta NOT pi_ref!

**Comment by liranringel** [comment]
@Paulmzr can you show the correct way?
